tutorial/other_packages/prebuilt_libraries (#2603)
conan-io/docs: tutorial/creating_packages/other_types_of_packages/package_prebuilt_binaries.rst

Package prebuilt binaries
=========================

There are specific scenarios in which it is necessary to create packages from existing binaries, for example from 3rd
parties or binaries previously built by another process or team that is not using Conan. Under these circumstances,
building from sources is not what you want.

You can package the local files in the following scenarios:

 1. When you are developing your package locally and you want to quickly create a package with the built artifacts, but as you don't want to rebuild again (clean copy) your artifacts, you don't want to call
    :command:`conan create`. This method will keep your local project build if you are using an IDE.
 2. When you cannot build the packages from sources (when only pre-built binaries are available) and you have them
    in a local directory.
 3. Same as 2 but you have the precompiled libraries in a remote repository.

Locally building binaries
-------------------------

> Use the :command:`conan new` command to create a "Hello World" C++ library example project:

Author: Actually, it works perfectly. But it is not the best to explain what is happening IMO

> Use the :command:`conan new` command to create a "Hello World" C++ library example project:

Author: Changed the explanations about the alternative/generic package method. Not sure what more Can i improve.

.. code-block:: bash

    $ conan new cmake_lib -d name=hello -d version=1.0

This will create a Conan package project with the following structure.

.. code-block:: text

  .
  ├── CMakeLists.txt
  ├── conanfile.py
  ├── include
  │   └── hello.h
  ├── src
  │   └── hello.cpp
  └── test_package
      ├── CMakeLists.txt
      ├── conanfile.py
      └── src
          └── example.cpp

We have a ``CMakeLists.txt`` file in the root, an ``src`` folder with the ``cpp`` files and, an ``include``
folder for the headers.

They also have a ``test_package/`` folder to test that the exported package is working correctly.

Now, for every different configuration (different compilers, architectures, build_type...):

1. We call :command:`conan install` to generate the ``conan_toolchain.cmake`` file and the ``CMakeUserPresets.json``
   that can be used in our IDE or calling CMake (only >= 3.23).

   .. code-block:: bash

       $ conan install . -s build_type=Release

2. We build our project calling CMake, our IDE, ... etc:

>    - *Traditional way*:

Reviewer: I woul just give one of the two alternatives, as this is not the focus of the tutorial and makes it a bit more confusing. You could choose for example the presets one and just mention that you could do it without presets or the opposite, but just having one code snippet.

              copy(self, "*.lib", local_lib_folder, os.path.join(self.package_folder, "lib"), keep_path=False)
              copy(self, "*.a", local_lib_folder, os.path.join(self.package_folder, "lib"), keep_path=False)

          def package_info(self):
              self.cpp_info.libs = ["hello"]

As we are developing a ``CMake`` project we declare a ``layout()`` method calling ``cmake_layout(self)``.
This will help locate the headers and the built libraries to be packaged.

The ``package()`` method is copying artifacts from the following directories that, thanks to the layout(), will always
point to the correct places:

- **os.path.join(self.source_folder, self.cpp.source.includedirs[0])** will always point to our local include folder.
- **os.path.join(self.build_folder, self.cpp.build.libdirs[0])** will always point to the location of the libraries when
  they are built, no matter if using a single-config CMake Generator or a multi-config one.

Now, for every different configuration (different compilers, architectures, build_type...):

1. We call :command:`conan install` to generate the ``conan_toolchain.cmake`` file and the ``CMakeUserPresets.json``

>    that we can be used in our IDE or calling CMake (only >= 3.23).

Reviewer: ```suggestion
   that can be used in our IDE or calling CMake (only >= 3.23).
```

Package prebuilt binaries
=========================

There are specific scenarios in which it is necessary to create packages from existing binaries, for example from 3rd
parties or binaries previously built by another process or team that is not using Conan. Under these circumstances,
building from sources is not what you want.

You can package the local files in the following scenarios:

>  1. When you are developing your package locally and you want to export the built artifacts to the local
>     cache. As you don't want to rebuild again (clean copy) your artifacts, you don't want to call
>     :command:`conan create`. This method will keep your build cache if you are using an IDE.

Reviewer 2: ```suggestion
 1. When you are developing your package locally and you want to quickly create a package with the built artifacts, but as you don't want to rebuild again (clean copy) your artifacts, you don't want to call
    :command:`conan create`. This method will keep your local project build if you are using an IDE.
```

 2. When you cannot build the packages from sources (when only pre-built binaries are available) and you have them
    in a local directory.
 3. Same as 2 but you have the precompiled libraries in a remote repository.

Locally building binaries
-------------------------

This is an example of scenario 1 explained in the introduction. In this case, you can run :command:`conan export-pkg`
command directly.

Please, first clone the sources to recreate this project. You can find them in the

> `examples2.0 repository <https://github.com/conan-io/examples2>`_ on GitHub:

Reviewer 2: Isn't this example reproducible from the ``conan new`` template? In theory it should, and things would be a bit simpler?

Author: If you change the `conanfile.py` it is.

Author: I've changed it. Please review
